src/shared/pam-util.c

```diff
@@ -21,6 +22,22 @@ int pam_syslog_errno(pam_handle_t *handle, int level, int error, const char *for
         return PAM_BUF_ERR;
 }
 
+int pam_syslog_pam_error(pam_handle_t *handle, int level, int error, const char *format) {
+        /* This wraps pam_syslog() but supports specifying %m at the very end (only!)
+         * to be replaced with pam_strerror(). */
+
+        if (endswith(format, "%m")) {
+                const char *t = strndupa_safe(format, strlen(format) - 2);
+                assert(!strchr(t, '%'));
+                pam_syslog(handle, level, "%s%s", t, pam_strerror(handle, error));
+        } else {
+                assert(!strchr(format, '%'));
+                pam_syslog(handle, level, "%s", format);
+        }
```

Comment:
hmm, instead of this magic behaviour, why not just do this as a pam_syslog_perror() helper instead, i.e. automatically suffix `": "` and the PAM error? I think this would be much less surprising and safer, and would be along the lines of classic UNIX apis.

Reply:
Because we have places where the `%m` is used, and places where it's not used. I think the helper is _fine_, i.e. it's a bit magic, but also the pam code is isolated and I think it's fine to use some custom helpers there. But if there's consensus that those helpers are too much, I can just drop them and use STRERROR() everywhere. The code is more verbose and possibly slightly less efficient, but it'd avoid all the magic then.

Reply:
i think there's no value in optimizing error paths. but i don't grok what you mean by using STRERROR() everywhere? that doesn't work for pam errors after all...

Reply:
what about this idea: don#t try to overload %m here for PAM errors. Instead, just add some super simple template logic: if `@pamerr@` or so is used in the string, replace it with the PAM error, and then leave `%m` to be the libc error as usual